### PR TITLE
[WIP] [toml/en] Add TOML support

### PR DESCRIPTION
**Pull request by aloisdg**
- [x] PR touches only one file (or a set of logically related files with similar changes made)
- [x] YAML Frontmatter formatted according to [CONTRIBUTING.md](https://github.com/adambard/learnxinyminutes-docs/blob/master/CONTRIBUTING.markdown)
- [x] Content changes are aimed at *intermediate to experienced programmers* (this is a poor format for explaining fundamental programming concepts) (I may improve it later)
- [ ] Seriously, look at it now. Watch for quotes and double-check field names.

-

Hello,

A new PR, for a new format I started to learn. TOML is an alternative to YAML or JSON. Looks the [repository](https://github.com/toml-lang/toml) for more informations. TOML is not stable yet, but it can be nice to have it if people still look to use it.

The PR is still in work in progress. I will fill it all along my learning. Feel free to comment it.

Cheers,

### CODE REVIEW

**Comment by aloisdg** [comment]
I think I made a quite nice overview of TOML. Now, it is time for review. I will start myself, of course, but feel free to step by and comment. Any help is welcome!

File: toml.html.markdown [hunk 1]
```diff
@@ -0,0 +1,251 @@
+---
+language: toml
+filename: learntoml.toml
+contributors:
+  - ["Alois de Gouvello", "https://github.com/aloisdg"]
+---
+
+TOML stands for Tom's Obvious, Minimal Language. It is a data serialisation language designed to be a minimal configuration file format that's easy to read due to obvious semantics.
+
+It is an alternative of YAML and JSON. It aims to be more human friendly than JSON and simpler that YAML. TOML is designed to map unambiguously to a hash table. TOML should be easy to parse into data structures in a wide variety of languages.
```
Comment (vendethiel) [inline toml.html.markdown]: an alternative **to**

File: toml.html.markdown [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+Be warned, TOML's spec is still changing a lot. Until it's marked as 1.0, you
+should assume that it is unstable and act accordingly. This document follow TOML v0.4.0.
```
Comment (vendethiel) [inline toml.html.markdown]: follows

File: toml.html.markdown [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+# A bare key must be non-empty, but an empty quoted key is allowed
+"" = "blank"     # VALID but discouraged
+'' = 'blank'     # VALID but discouraged
+
+##########
+# String #
+##########
+
+# All strings must contain only valid UTF-8 characters.
+# We can escape and some characters have a compact escape sequence.
```
Comment (vendethiel) [inline toml.html.markdown]: "We can escape characters and some of them"?

File: toml.html.markdown [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+# For example, \t add a tabulation. Refers to spec to get all of them.
```
Comment (vendethiel) [inline toml.html.markdown]: is a

File: toml.html.markdown [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+[[products]]
+
+[[products]]
+name = "Nail"
+sku = 284758393
+color = "gray"
+
+# You can create nested arrays of tables as well. Each double-bracketed
+# sub-table will belong to the nearest table element above it.
```
Comment (vendethiel) [inline toml.html.markdown]: I didn't get that part.

File: toml.html.markdown (hunk 4 above)
Comment (aloisdg) [inline toml.html.markdown]: Thanks!

File: toml.html.markdown (hunk 2 above)
Comment (aloisdg) [inline toml.html.markdown]: "Follows" is fine, isn'it?

File: toml.html.markdown (hunk 3 above)
Comment (aloisdg) [inline toml.html.markdown]: Better indeed :)

File: toml.html.markdown (hunk 2 above)
Comment (vendethiel) [inline toml.html.markdown]: sure :-).

File: toml.html.markdown (hunk 5 above)
Comment (aloisdg) [inline toml.html.markdown]: The part ahead or below?

Comment (vendethiel) [inline toml.html.markdown]: Well, the section about sub-tables

File: toml.html.markdown [hunk 6]
```diff
@@ -228,22 +228,47 @@ color = "gray"
...
   name = "banana"
 
-  [[fruit.variety]]
-    name = "plantain"
+  [[fruit.color]]
+    name = "yellow"
+    note = "I am an array's item too but banana's one"
+```
 
+In JSON land, this code will be:
```
Comment (aloisdg) [inline toml.html.markdown]: @vendethiel Hello, I am not sure about this one. Am I allowed to insert a json snippet in my file? Should I move it insde the toml snippet as a multi-lines comment?

File: toml.html.markdown (hunk 5 above)
Comment (aloisdg) [inline toml.html.markdown]: I rewrote it :)

File: toml.html.markdown (hunk 6 above)
Comment (vendethiel) [inline toml.html.markdown]: Seems just fine to me.

Comment (aloisdg) [inline toml.html.markdown]: Nice! 👍

**Comment by aloisdg** [comment]
I share this post on the TOML repository. i hope to get some feedback from them. I think we are near the end. 😄

File: toml.html.markdown [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+# For example, \t add a tabulation. Refers to the spec to get all of them.
+basicString = "are surrounded by quotation marks. \"I'm quotable\". Name\tJos"
+
+multiLineString = """
+are surrounded by three quotation marks
+on each side and allow newlines."""
+
+literalString = 'are surrounded by single quotes. Escaping are not allowed.'
+
+MultiLineLiteralString = """
```
Comment (mdickie) [inline toml.html.markdown]: Shouldn't MultiLineLiteralString start with ''' instead of """

Comment (aloisdg) [inline toml.html.markdown]: @mdickie No MultiLineString (and MultiLineLiteralString) uses `"""` (3 `"`). See [spec](https://github.com/toml-lang/toml#string) for more informations 😄

Comment (mdickie) [inline toml.html.markdown]: According to spec:
 > Multi-line literal strings are surrounded by three single quotes on each side and allow newlines.

Comment (aloisdg) [inline toml.html.markdown]: Oh! You are right! I mixed both. I am fixing it. Thank you!

File: toml.html.markdown [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+
+In JSON land, this code will be:
+
+```json
+{
+  "fruit": [
+    {
+      "name": "apple",
+      "geometry": { "shape": "round", "note": "..."},
+      "variety": [
```
Comment (robochat) [inline toml.html.markdown]: this should probably be "color" rather than "variety".

File: toml.html.markdown [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+        { "name": "red", "note": "..." },
+        { "name": "green", "note": "..." }
+      ]
+    },
+    {
+      "name": "banana",
+      "variety": [
```
Comment (robochat) [inline toml.html.markdown]: this should probably be "color" rather than "variety".

File: toml.html.markdown (hunk 8 above)
Comment (aloisdg) [inline toml.html.markdown]: Indeed!

File: toml.html.markdown (hunk 9 above)
Comment (aloisdg) [inline toml.html.markdown]: Yup

**Comment by aloisdg** [comment]
Any updates on this? I think we can ship it.

**Comment by vendethiel** [comment]
Nice, thanks! :ship:

**Pull request merged**